Index: linux-3.10.14/include/net/ip.h
===================================================================
--- linux-3.10.14.orig/include/net/ip.h	2016-01-08 14:06:57.000000000 +0800
+++ linux-3.10.14/include/net/ip.h	2017-07-25 16:46:27.009168800 +0800
@@ -396,6 +396,7 @@
 	IP_DEFRAG_VS_FWD,
 	IP_DEFRAG_AF_PACKET,
 	IP_DEFRAG_MACVLAN,
+	IP_DEFRAG_IP4OV6JP,	//Add for support cutomized IPv4 over IPv6
 };
 
 int ip_defrag(struct sk_buff *skb, u32 user);
